src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -934,6 +1116,205 @@ - (void)readAttributesWithEndpointID:(NSNumber * _Nullable)endpointID
+                             dispatch_async(queue, ^{
+                                 reportHandler(@[ @ {
+                                     MTRAttributePathKey : [[MTRAttributePath alloc] initWithPath:pathCopy],
+                                     MTRDataKey : valueObject
+                                 } ],
+                                     nil);
+                             });
+                         };
+
+                   auto onEventReportCb = [queue, reportHandler](const app::ConcreteClusterPath & clusterPath,
+                                              const EventId aEventId, const MTRDataValueDictionaryDecodableType & data) {
+                       id valueObject = data.GetDecodedObject();
+                       app::ConcreteEventPath pathCopy(clusterPath.mEndpointId, clusterPath.mClusterId, aEventId);
+                       dispatch_async(queue, ^{
+                           reportHandler(
+                               @[ @ { MTREventPathKey : [[MTREventPath alloc] initWithPath:pathCopy], MTRDataKey : valueObject } ],
+                               nil);
+                       });
+                   };
+
+                   auto establishedOrFailed = chip::Platform::MakeShared<BOOL>(NO);
+                   auto onFailureCb = [establishedOrFailed, queue, subscriptionEstablished, reportHandler](
+                                          const app::ConcreteClusterPath * clusterPath, const uint32_t aValueId, CHIP_ERROR error) {
+                       if (!(*establishedOrFailed)) {
+                           *establishedOrFailed = YES;
+                           if (subscriptionEstablished) {
+                               dispatch_async(queue, subscriptionEstablished);
+                           }
+                       }
+                       if (reportHandler) {
+                           dispatch_async(queue, ^{
+                               reportHandler(nil, [MTRError errorForCHIPErrorCode:error]);
+                           });
+                       }
+                   };
+
+                   auto onEstablishedCb = [establishedOrFailed, queue, subscriptionEstablished]() {
+                       if (*establishedOrFailed) {
+                           return;
+                       }
+                       *establishedOrFailed = YES;
+                       if (subscriptionEstablished) {
+                           dispatch_async(queue, subscriptionEstablished);
+                       }
+                   };
+
+                   MTRReadClientContainer * container = [[MTRReadClientContainer alloc] init];
+                   container.deviceID = self.nodeID;
+
+                   std::vector<app::AttributePathParams> attributePathParamsList;
+                   std::vector<app::EventPathParams> eventPathParamsList;
+
+                   if (attributePaths != nil) {
+                       for (uint8_t i = 0; i < [attributePaths count]; i++) {
```

Comment:
See comments in the read version of this code.  Seems like we might want to have some static helper functions for "convert this NSArray<MTRAttributeRequestPath> to std::vector<AttributePathParams>" and similar for the event case...  But see below.

Reply:
modified using array memory allocation.